### PR TITLE
Fix Dancer order & add some Dancer tests

### PR DESCRIPTION
**Pull request by MacChaeger**
I only added these particular tests so that I could make sure this code worked without having to run a server and do a mock battle 30 times, so if you want me to add more or remove any of them, I can.

P.S. I don't think anyone's checked whether or not Dancer has the same glitch as Trick Room with Pokemon over 1809 (1808?) speed.

### CODE REVIEW

File: data/scripts.js [hunk 1]
```diff
@@ -88,6 +88,31 @@ exports.BattleScripts = {
...
+			for (const side of this.sides) {
+				for (const slot of side.active) {
+					let currentPoke = slot;
+					if (!currentPoke || !currentPoke.hp || pokemon === currentPoke || move.isExternal) continue;
+					if (currentPoke.ability === 'dancer') {
+						dancers.push(currentPoke);
+					}
+				}
+			}
+			// Dancer activates in reverse speed order with or without trick room, so we simulate this by storing the current Trick Room status and restoring it after sorting by speed
```
Comment (Zarel) [inline data/scripts.js]: ...You can implement a Dancer pseudoweather, you know.

Comment (Zarel) [reply data/scripts.js]: Instead of overloading Trick Room.

**Comment by Marty-D** [comment]
Apparently I need to audit every claim made in the research thread since this isn't true either. Dancer activation order is the opponent's left, then right, then the user's side, regardless of Speed.

**Comment by Zarel** [comment]
@DaWoblefet?

**Comment by MacChaeger** [comment]
I can't watch the battle video @DaWoblefet made because my 3DS doesn't connect to my campus's wifi, but I tested it just now (for the fourth time) against some wild Oricorio and both the left and right slots moved first with different Oricorio.

**Comment by MacChaeger** [comment]
When you say "Opponent's left" that'd be to the user's right, correct?

**Comment by Marty-D** [comment]
>When you say "Opponent's left" that'd be to the user's right, correct?

Yeah, and you should basically never test anything in wild battles since plenty of stuff behaves differently. In this case, an SOS battle is not a true Double Battle, so the concept of opponent and left/right slots doesn't apply.

**Comment by Marty-D** [comment]
Thinking about it some more, I notice both the mock battle and Trainer Double Battle I usually test with have the opponent's left as the slower of the two, and despite giving that one Dancer first and letting it raise its Speed beyond its ally, it still activated Dancer first. After testing again with a different mock battle, it's clear that what's actually happening is Dancer ignores actual Speed and just uses initial Speed to sort slowest to fastest.

**Comment by MacChaeger** [comment]
Is there a tie-breaker after how long they've had the ability?

**Comment by Marty-D** [comment]
>Is there a tie-breaker after how long they've had the ability?

It's not possible to tie "having an Ability for X amount of time", since it starts the instant a Pokemon becomes active. Even on the first turn, the player who accepted the challenge sends their Pokemon out first, and each player's left slot will have "had its Ability longer" than the right.

**Comment by MacChaeger** [comment]
Ok so just using `abilityOrder` as the final tie-breaker is sufficient?

**Comment by Marty-D** [comment]
Yeah, just use the inverse of whatever Lightning Rod is doing for its ties.

**Comment by MacChaeger** [comment]
Ok, thanks, that's what I was thinking.

File: data/scripts.js [hunk 2]
```diff
@@ -88,6 +88,28 @@ exports.BattleScripts = {
 		this.useMove(baseMove, pokemon, target, sourceEffect, zMove);
 		this.singleEvent('AfterMove', move, null, pokemon, target, move);
 		this.runEvent('AfterMove', pokemon, target, move);
+
+		// Dancer's activation order is completely different from any other event, so it's handled separately
+		if (move.flags['dance']) {
+			let dancers = [];
+			for (const side of this.sides) {
+				for (const slot of side.active) {
+					let currentPoke = slot;
```
Comment (urkerab) [inline data/scripts.js]: Why not just `for (const currentPoke of side.active) {`?

File: data/scripts.js [hunk 3]
```diff
@@ -88,6 +88,28 @@ exports.BattleScripts = {
...
+					if (!currentPoke || !currentPoke.hp || pokemon === currentPoke || move.isExternal) continue;
```
Comment (urkerab) [inline data/scripts.js]: `move.isExternal` should probably be tested earlier along with `move.flags['dance']`.

File: data/scripts.js [hunk 4]
```diff
@@ -88,6 +88,28 @@ exports.BattleScripts = {
...
+					if (currentPoke.ability === 'dancer') {
```
Comment (urkerab) [inline data/scripts.js]: This needs to be `if (currentPoke.hasAbility('dancer')) {` in case of Gastro Acid/Core Enforcer.

**Comment by Marty-D** [comment]
Looks good, tested the same things I did in-game and got the same results. Just address the stuff urkerab mentioned and I think we're good to go here. Nicely done! 👍

File: data/scripts.js (hunk 2 above)
Comment (Zarel) [reply data/scripts.js]: Yeah, `slot` is not actually something you want to call a Pokémon.

File: data/scripts.js (hunk 3 above)
Comment (Zarel) [reply data/scripts.js]: What's `isExternal` again?

Comment (Marty-D) [reply data/scripts.js]: Whether the move was used by a Dancer activation or not.

File: data/scripts.js (hunk 2 above)
Comment (MacChaeger) [reply data/scripts.js]: Oh yeah that was from before when I was using for...length and `slot` was `this.battle.sides[i].active[j]`

**Pull request merged**